File: app/src/main/java/at/shockbytes/dante/util/backup/BackupManager.java
package at.shockbytes.dante.util.backup;

import android.support.v4.app.FragmentActivity;

import java.util.List;

import at.shockbytes.dante.util.books.Book;
import at.shockbytes.dante.util.books.BookManager;
import io.reactivex.Completable;
import io.reactivex.Observable;

/**
 * @author Martin Macheiner
 *         Date: 30.04.2017.
 */

public interface BackupManager {

    interface OnConnectionStatusListener {

        void onConnectionFailed();

        void onConnected();
    }

    int RESOLVE_CONNECTION_REQUEST_CODE = 0x1231;

    enum RestoreStrategy {MERGE, OVERWRITE}

    void connect(FragmentActivity activity, OnConnectionStatusListener onStatusListener);

    void disconnect();

    boolean isAutoBackupEnabled();

    void setAutoBackupEnabled(boolean autoBackupEnabled);

    long getLastBackupTime();

    Observable<List<BackupEntry>> getBackupList();

    Completable removeBackupEntry(BackupEntry entry);

    Completable removeAllBackupEntries();

    Completable backup(List<Book> books);

    Completable restoreBackup(FragmentActivity activity, BackupEntry entry,
                                   BookManager bookManager, RestoreStrategy strategy);

}

File: app/src/main/java/at/shockbytes/dante/backup/BackupManager.kt
package at.shockbytes.dante.backup

import androidx.fragment.app.FragmentActivity
import at.shockbytes.dante.backup.model.BackupMetadata
import at.shockbytes.dante.backup.model.BackupServiceConnectionException
import at.shockbytes.dante.util.RestoreStrategy
import at.shockbytes.dante.core.book.BookEntity
import at.shockbytes.dante.core.data.BookEntityDao
import io.reactivex.Completable
import io.reactivex.Single

/**
 * @author Martin Macheiner
 * Date: 30.04.2017.
 */
@Deprecated("Use BackupRepository instead")
interface BackupManager {

    var lastBackupTime: Long

    val backupList: Single<List<BackupMetadata>>

    @Throws(BackupServiceConnectionException::class)
    fun connect(activity: FragmentActivity): Completable

    fun close(books: List<BookEntity>? = null)

    fun removeBackupEntry(entry: BackupMetadata): Completable

    fun removeAllBackupEntries(): Completable

    fun backup(books: List<BookEntity>): Completable

    fun restoreBackup(
        entry: BackupMetadata,
        bookDao: BookEntityDao,
        strategy: RestoreStrategy
    ): Completable
}
